### ISSUE TITLE
HTML codes in title not converted / corrected on API and NZB

### ISSUE THREAD

**Issue by Mirx** [spotweb/spotweb#129]
Too start, it is a very limited issue. So if it's not a simple fix, please close the issue.

Some posters have the HTML codes in their spot titles instead of the ASCII characters.

In the spots overview, the title is corrected to the respective ASCII character.
However, when downloading an NZB file, or requesting an RSS feed through the API, the original HTML code is presented in the title.
Due to the HTML code in the names in the RSS feeds, automated downloaders like Sonarr are not able to parse the output correctly and match it to a series.

These posts can be easily found by the following SQL query:
SELECT * FROM `spots` WHERE `title` LIKE '%&#%'

For example:
Database - Spotoverview - API - NZB (Windows)
`Lee&#39;s - Lee's - Lee&#39;s - Lee_#39_s`

Reference I used for the HTML codes in relation to their ascii counterparts.
http://www.ascii.cl/htmlcodes.htm

**Comment by Sweepr** [comment]
I ran the query you provided and i'm only able to find 9 spots, 7 of them are of the same poster, so i suspect he's was/is using outdated software to post, or something is messing up copy/pasting.

**Comment by mesa57** [comment]
However, in the spotweb overview and also in spotnet they are showing up normal.
In my opinion this is a (simple to solve) bug in the api.

**Comment by Mirx** [comment]
Noticed something else in tinkering with it. I figured, let's just change the title in the database. 

When I do that, and alter the entry from `Lee&#39;s to Lee's` the API output still shows a HTML code. It did however change from `Stan Lee&#39;s` to `Stan Lee&#039;s` 

Then looking at posts with generic ' in the name, like `MARVEL's AGENT CARTER  S02E04 H.264 (WEB-DL) DD5.1 1080p -Q o Q-` I found they are also outputted with HTML code on the API.

`MARVEL&#039;s AGENT CARTER S02E04 H.264 (WEB-DL) DD5.1 1080p -Q o Q-`

So it seems a bit bigger then just the posts that have the HTML code in the title.

**Comment by mesa57** [comment]
Spotweb uses the function htmlspecialchars to encode special chars in xml, which is correct.
However, option ENT_QUOTES decodes single quotes into the &#39 
This is not necessary on XML content, only on XML attributes. 
So changing that options to ENT_COMPATS helps. 
But not for titles where &#39 already is in the (database) title. 
What helps is first to a htmlspecialchars_decode and then a htmlspecialchars to encode again.

I will make a change and submit a pull request, after I have tested also with json output.

**Comment by mesa57** [comment]
A little update: I discovered that spotweb initialy stores the title correct into the spots table with no html encoding for single quotes in the title.
However, if you open a spot, it looks like it updates the title in the spot table with the content of <title> ...</title> from the spot's xml.
And that in that content, single quotes are (often) encoded as &#39

Reason for this update is unclear for me.

**Comment by Mirx** [comment]
Good catch. Also, downloading the nzb without opening the spot has the similar behaviour.

**Comment by mesa57** [comment]
For both, downloading of the fullspot is necessary, so I guess it happens there.

**Comment by Mirx** [comment]
The overwrite happens in line 442 of spotweb/lib/services/Retriever/Services_Retriever_Spots.php

**Comment by Mirx** [comment]
Edit, my bad…

spotweb/lib/dao/Base/Dao_Base_Spot.php line 452 is the actual replace. Turn out there are more places doing something similar.

As the spotsfull actually contains the html codes, it seems that should be the source to “fix”.

To me, it seems weird to be encoding UTF-8 into UTF-8

In posting the it seems HTML codes are used.

/**
		 * If the post's character do not fit into ISO-8859-1, we HTML
		 * encode the UTF-8 characters so we can properly post the spots
		 */
		if (mb_detect_encoding($spot['title'], 'UTF-8, ISO-8859-1', true) == 'UTF-8') {
			$spot['title'] = mb_convert_encoding($spot['title'], 'HTML-ENTITIES', 'UTF-8');

So shouldn’t we be converting the HTML entities back into UTF-8 when storing the title in the spotfull database?

**Comment by mesa57** [comment]
Spot titles may (not necesary) be encoded in UTF-8
I do not know if other spotnet client can handle that.

**Comment by Mirx** [comment]
Let me rephrase, as I have fallen into the trap of thinking that I understood it and might have asked something stupid ;)

$spot['title'] seems to be stored  in the database without HTML code. 
$fullSpot['title'] seems to be stored in the database with HTML code.

As both seem to be stored as UTF-8, why is the second one preferred?

**Comment by mesa57** [comment]
The original message from the usenet server contains : 
- No html encoded characters (at least not single quotes) in the Subject: line -> which goes to spots.title
- Html encoded characters in the fullspot XML <title>  .... </title> also for single quotes --> which goes to fullspots.title

**Comment by Mirx** [comment]
I would have expected 

changing
        $fullSpot['title'] = mb_convert_encoding($fullSpot['title'], 'UTF-8', 'UTF-8');
to
        $fullSpot['title'] = mb_convert_encoding($fullSpot['title'], 'UTF-8', 'HTML-ENTITIES');

would revert DC&#39;s back to DC's

But I might just have been doing that in the wrong section of the code.

**Comment by mesa57** [comment]
The conversion from UTF-8 only takes place when a <title> is UTF-8 encoded. I think most of them are not.

**Comment by mesa57** [comment]
You can try the solution I commited on my repository mesa57/spotweb
If tested ok, I will issue a pull request to merge it to the master branch of spotweb/spotweb

**Comment by Mirx** [comment]
Will do. Hopefully I'll have some time to make the changes on sunday, then it might take some time before a testable post is available.

**Comment by mesa57** [comment]
Just put DC's Legends of Tomorrow in you're sonarr :wink:

**Comment by Mirx** [comment]
I have, and a bunch of others :)
Just did a dry run and noticed the following behaviour:
* Database entries aren't overwritten any more with the HTML tags
* Spotweb and API do not contain any HTML tags, so release are found correctly
* If grabbed automatically by Sonarr, Sonarr uses its name instead of the NZB name, and downloads are succesfull

```
16-3-19 08:57:54.1|Debug|Parser|Parsing string 'Stan Lee's Lucky Man S01E09 1080p HDTV X264 TLA'
16-3-19 08:57:54.1|Debug|Parser|Episode Parsed. Stan Lee's Lucky Man - S01E09 
```

* The NZB file names are not generated identically to the spot name. Lee's -> Lee_s etc. So manually adding a NZB file is still a no-go area, as Sonarr won't parse the series title correctly.

```
16-3-19 08:55:24.6|Debug|Parser|Parsing string 'Stan Lee_s Lucky Man S01E09 1080p HDTV X264 TLA'
16-3-19 08:55:24.6|Debug|Parser|Episode Parsed. Stan Lee s Lucky Man - S01E09 
```

**Comment by mesa57** [comment]
I hesitated to change the NZB filename. In linux it is officialy not allowed to have quotes in a filename. That's why they are replaced by _

**Comment by Mirx** [comment]
I knew they could be a hassle, didn't know that they weren't allowed (as they are escapable).

Directory
`cd Stan\ Lee\'s\ Lucky\ Man/`
Filename
`cp Stan\ Lee\'s\ Lucky\ Man\ S01E0`

**Comment by mesa57** [comment]
It is conform the posix standard : 
http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_276

That doens't mean that certain os's allow use of other characters, but spotweb uses the POSIX standard. That's why I am hestitating to change that. In fact, Sonarr should support the POSIX standard when searching for filenames.

**Comment by Mirx** [comment]
Sonarr does work, even on searches, as that uses the API.
It just when I manually click the NZB button in Spotweb, it get's sent to the usenet client with the NZB name. Same as I would manually add the NZB.
Sonarr then fails the parse when reading the usenet client history, as the NZB name is used.

I wasn't saying you were wrong, just saying I never knew that the limitation existed as I'm a complete *nix n00b.

So if there's is a risk involved, don't do it.

**Comment by mesa57** [comment]
Sonarr should be aware that the filenames he is searching for should conform to the Sonarr standard.
If you agree, I will merge this change with the master branch ?

**Comment by Mirx** [comment]
I tested about 5 of 6 previously problematic shows by manually searching the RSS feeds. All results seem fine.

Hell's Kitchen used to be the worst, because there was also a "-" which got converted. Seems to work fine now. 

I can't test automatic grabs, first problematic show does not air until wednesday. But all the behaviour makes me pretty confident it will work without problems.

So I would say, job well done, and merge with the master.

**Comment by mesa57** [comment]
My sickbeard is doing grab's for a week now, including DC's 
I will merge the pull request and close the issue.
